src/mscorlib/shared/System/Memory.cs

```diff
@@ -0,0 +1,313 @@
+                }
+                else
+                {
+                    var handle = GCHandle.Alloc(Unsafe.As<T[]>(_arrayOrOwnedMemory), GCHandleType.Pinned);
+                    void* pointer = Unsafe.Add<T>((void*)handle.AddrOfPinnedObject(), _index);
+                    memoryHandle = new MemoryHandle(null, pointer, handle);
+                }
+            }
+            else
+            {
+                if (_index < 0)
+                {
+                    Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory).Retain();
+                    memoryHandle = new MemoryHandle(Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory));
+                }
+                else
+                {
+                    memoryHandle = new MemoryHandle(null);
+                }
+            }
+            return memoryHandle;
+        }
+
+        /// <summary>
+        /// Get an array segment from the underlying memory. 
+        /// If unable to get the array segment, return false with a default array segment.
+        /// </summary>
+        public bool TryGetArray(out ArraySegment<T> arraySegment)
+        {
+            if (_index < 0)
+            {
+                if (Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory).TryGetArray(out var segment))
+                {
+                    arraySegment = new ArraySegment<T>(segment.Array, segment.Offset + (_index & bitMask), _length);
+                    return true;
+                }
+            }
+            else
+            {
+                arraySegment = new ArraySegment<T>(Unsafe.As<T[]>(_arrayOrOwnedMemory), _index, _length);
+                return true;
+            }
+
+            arraySegment = default(ArraySegment<T>);
+            return false;
+        }
+
+        /// <summary>
+        /// Copies the contents of this span from the memory into a new array.  This heap
+        /// allocates, so should generally be avoided, however it is sometimes
+        /// necessary to bridge the gap with APIs written in terms of arrays.
+        /// </summary>
+        public T[] ToArray() => Span.ToArray();
+
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public override bool Equals(object obj)
+        {
+            if (obj is ReadOnlyMemory<T> readOnlyMemory)
+            {
+                return readOnlyMemory.Equals(this);
+            }
+            else if (obj is Memory<T> memory)
+            {
+                return Equals(memory);
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Returns true if the memory points to the same array and has the same length.  Note that
+        /// this does *not* check to see if the *contents* are equal.
+        /// </summary>
+        public bool Equals(Memory<T> other)
+        {
+            return
+                _arrayOrOwnedMemory == other._arrayOrOwnedMemory &&
+                (_index & bitMask) == (other._index & bitMask) &&
```

Comment:
Why is the bit masking necessary?

Reply:
I was thinking of something like this:
```C#
    var array = new byte[100];
    var memory = new Memory<byte>(array);
    var ownedMemory = (OwnedMemory<byte>)array;
    var memoryOwned = ownedMemory.Memory;
    Assert.False(memory .Equals(memoryOwned)); // This will succeed even without the bit masking 
```

The bit masking isn't necessary. The first equality catches if the object is an array or an OwnedMemory.